Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm

```diff
@@ -40,10 +40,223 @@
+    };
+
+    auto updateURL = [&](CompletionHandler<void(WebExtensionTab::Error)>&& stepCompletionHandler) {
+        if (!parameters.url) {
+            stepCompletionHandler(std::nullopt);
+            return;
+        }
+
+        tab->loadURL(parameters.url.value(), WTFMove(stepCompletionHandler));
+    };
+
+    auto updatePinned = [&](CompletionHandler<void(WebExtensionTab::Error)>&& stepCompletionHandler) {
+        if (!parameters.pinned || parameters.pinned.value() == tab->isPinned()) {
+            stepCompletionHandler(std::nullopt);
+            return;
+        }
+
+        if (parameters.pinned.value())
+            tab->pin(WTFMove(stepCompletionHandler));
+        else
+            tab->unpin(WTFMove(stepCompletionHandler));
+    };
+
+    auto updateMuted = [&](CompletionHandler<void(WebExtensionTab::Error)>&& stepCompletionHandler) {
+        if (!parameters.muted || parameters.muted.value() == tab->isMuted()) {
+            stepCompletionHandler(std::nullopt);
+            return;
+        }
+
+        if (parameters.muted.value())
+            tab->mute(WTFMove(stepCompletionHandler));
+        else
+            tab->unmute(WTFMove(stepCompletionHandler));
+    };
+
+    auto updateParentTab = [&](CompletionHandler<void(WebExtensionTab::Error)>&& stepCompletionHandler) {
+        auto currentParentTab = tab->parentTab();
+        auto newParentTab = parameters.parentTabIdentifier ? getTab(parameters.parentTabIdentifier.value()) : nullptr;
+
+        if (currentParentTab == newParentTab) {
+            stepCompletionHandler(std::nullopt);
+            return;
+        }
+
+        tab->setParentTab(newParentTab, WTFMove(stepCompletionHandler));
+    };
+
+    updateActiveAndSelected([&](WebExtensionTab::Error activeOrSelectedError) {
+        if (activeOrSelectedError) {
+            completionHandler(std::nullopt, activeOrSelectedError);
+            return;
+        }
+
+        updateURL([&](WebExtensionTab::Error urlError) {
+            if (urlError) {
+                completionHandler(std::nullopt, urlError);
+                return;
+            }
+
+            updatePinned([&](WebExtensionTab::Error pinnedError) {
+                if (pinnedError) {
+                    completionHandler(std::nullopt, pinnedError);
+                    return;
+                }
+
+                updateMuted([&](WebExtensionTab::Error mutedError) {
+                    if (mutedError) {
+                        completionHandler(std::nullopt, mutedError);
+                        return;
+                    }
+
+                    updateParentTab([&](WebExtensionTab::Error parentError) {
+                        if (parentError) {
+                            completionHandler(std::nullopt, parentError);
+                            return;
+                        }
+
+                        completionHandler(tab->parameters(), std::nullopt);
+                    });
+                });
+            });
+        });
+    });
```

Comment:
A bit of a shame this nesting is necessary to perform all of these steps, but can't think of anything better.

Reply:
Yeah… C++ needs `await` to solve that.